modules/telemetry/telemetry/examples/log_sink_example.cpp

```diff
@@ -0,0 +1,25 @@
+//=================================================================================================
+// Copyright (C) 2023-2024 HEPHAESTUS Contributors
+//=================================================================================================
+#include <memory>
+#include <string>
+
+#include "hephaestus/telemetry/log_sinks/absl_sink.h"
+#include "hephaestus/telemetry/struclog.h"
+#include "hephaestus/utils/stack_trace.h"
+
+namespace ht = heph::telemetry;
```

Comment:
I agree however in this case it would be:
```
heph::telemetry::log(heph::telemetry::LogEntry{ heph::telemetry::Level::WARN, "examples", "testing absl log sink" } | "num"_f(num));
```

In general I love the current interface, but it is really verbose, I am thinking how can we try to make it more compact.

Reply:
We break the convention of module names in the namespace for exceptions --> `heph::throwException()`. We could do the same here --> `heph::log()`. Wdyt?

Reply:
one variant would be 
´´´ht::log({Level::WARN, "examples", "testing absl log sink", "num"_f(num)});´´´
so we drop the pipe operator for a variadic constructor so that the compiler can deduce the LogEntry type. We could put the Level in the literals namespace since that one has to be used anyways for the literals. 

we also could introduce a literal for the level
´´´ht::log({"examples", "testing absl log sink"WARN, "num"_f(num)});´´´
since we could argue that the message itself is a warning or an error but that seems quite unusual and might be confusing.

Reply:
or we add overloads for log like ht::log_warn

Reply:
oh here we get to the meaty part. I love the design of the interface, but I think we need to also optimize for being compact (everything else I would object this, but here the log needs to be simple and frictionless). 
That said my 2 cents:
* Drop the `|` it was slick, but the advantage of being able to drop `LogEntry` outweigh it.
* I would keep the level as parameter instead of `log_warn` as it makes it more flexible and reduce a lot of boilerplate
  * This because I would also like to have `log_if` `log_every_n` ... so it would make too many combinations.
* I think Lorenz point of putting just `log` directly in `heph` namespace makes sense, exactly as we did for exceptions.

Reply:
if we want to emphasize readability on the message an interesting variant would also be a literal like
```cpp
ht::log("memory limit exceeded"msg(Level::Warn, "my module") | "mem"_f(memory_used) | "space"_f(freespace));
// or 
ht::log("memory limit exceeded"warn("my module") | "mem"_f(memory_used) | "space"_f(freespace));
```
I kind of like it since the main part, the error message is the first thing you read. but I can understand if this is too exotic and we stick with the variadic constructor.
@filippobrizzi @lievenvanderheide @lhruby thoughts?

Reply:
yeah I see that, I would not put it into the log header, but separately...  chrono and filesystem do it also with using namespace. I would think its fine, for source files I would even think its fine to use using top of the heph namespace, in headers we should advise to only use it in local scopes (like functions) as not to include it unconsciously.
if we do not want to add the line boilerplate everywhere, we could also put the literals in an additional header and add using namespace there. I do not see a case where you want to include the literals without adding using namespace.

Reply:
exactly, but in general I don't see a single use case where you want to use the log without the literals if we go for this final proposal. So what is the downside of adding `using namespace` in the header?

Reply:
nasty conflicts if someone decides to do the same for his or her library with same literal names...

Reply:
very verbose :)
```cpp
ht::log(ht::LogEntry(ht::Level::Warn,"my modue", "memory limit exceeded") | ht:.Field{.key="mem", .val=memory_used} | ht::Field{.key="space",.val=freespace});
```
dont get me wrong I am good either way, just want to raise that thought before going down that road.

Reply:
thanks for the recap!
* we also discussed the variadic constructor but at least @filippobrizzi and me like the fields operator version better :)
* for the namespace we proposed to put log into the main heph namespace

also we could put the `using namespace hephaestus::telemetry::log; `into the literals header since this is where the syntactic sugar lies to make the logging convenient

Reply:
The functionality for sure is cool but I have some concerns about the interface and usability. Its some what cumbersome to log in that way tbh. Would one way be that we first configure the logger on a binary basis and then use macros to automate some of the literals, modules etc? There is a reasons why most logging framworks use macros for that 😅

Reply:
@johannes-graeter I sync with @floriantschopp  on usability and we were thinking of something like this:
```
heph::log(heph::Level::Warn, "my message", {{"num", num}, {"speed", speed}});
```
* Drop my request of `module`. Given that we do not have a consistent way of setting it, it will create more damage than good, and we have the file location so we should be good.
* We can have an inline vector of tags, which makes the syntax even simpler

This should work, as the compiler should be able to infer the type, right? Uneles the problem is that Field itself is templated?
```
void log(Level level, std::string message, std::vector<Field> fields)
```

Reply:
> @johannes-graeter I sync with @floriantschopp  on usability and we were thinking of something like this:
> 
> ```
> 
> heph::log(heph::Level::Warn, "my message", {{"num", num}, {"speed", speed}});
> 
> ```
> 
> * Drop my request of `module`. Given that we do not have a consistent way of setting it, it will create more damage than good, and we have the file location so we should be good.
> 
> * We can have an inline vector of tags, which makes the syntax even simpler
> 
> 
> 
> This should work, as the compiler should be able to infer the type, right? Uneles the problem is that Field itself is templated?
> 
> ```
> 
> void log(Level level, std::string message, std::vector<Field> fields)
> 
> ```

Ok that's fine with me, outside of this context you really have to explain to me what a module is and how it is generated...

For the interface:
I guess we could have 
```cpp
struct Field {
Field(std::string key, auto val): key(key),val(std::to_string){}
std::string key;
std::string val;
};
```

We would need to move the quotable nonquotable logic as well to the field which I don't like too much...
What about the variadic interface 

```cpp
heph::log(heph::Level::Warn, "my message", {num", num}, {"speed", speed});
```

It's fairly standard interface for structures logging (at least from the libs I know in golang and rust) and visual separation is fine with the pair brackets.

Reply:
coverged to a suitable layout as described below